vllm/worker/model_runner.py

```diff
@@ -165,6 +165,287 @@ def from_broadcasted_tensor_dict(
+
+        # Prompt adapter data. Note that when is_prompt=True,
+        # we expect only one sequence in the group.
+        if self.enable_prompt_adapter:
+            prompt_adapter_id = seq_group_metadata.prompt_adapter_id
+            if prompt_adapter_id > 0 and is_prompt:
+                query_len = self.query_lens[-1]
+                self.prompt_adapter_requests.add(
+                    seq_group_metadata.prompt_adapter_request)
+
+                num_tokens = seq_group_metadata.\
+                    prompt_adapter_num_virtual_tokens
+                pm = [prompt_adapter_id
+                      ] * num_tokens + [0] * (query_len - num_tokens)
+                self.prompt_adapter_index_mapping += pm
+                self.prompt_adapter_prompt_mapping.extend(
+                    [prompt_adapter_id] *
+                    (query_len if seq_group_metadata.sampling_params
+                     and seq_group_metadata.sampling_params.prompt_logprobs
+                     else 1))
+
+        # Multi-modal data.
+        mm_data = seq_group_metadata.multi_modal_data
+        if mm_data:
+            mm_kwargs = self.multi_modal_input_mapper(mm_data)
+            self.multi_modal_inputs_list.append(mm_kwargs)
+
+    def build(self) -> ModelInputForGPU:
+        if not self.input_tokens:
+            return self.model_input_cls()
+
+        batch_size = len(self.input_tokens)
+        use_captured_graph = (
+            self.decode_only and not self.runner.model_config.enforce_eager
+            and batch_size <= _BATCH_SIZES_TO_CAPTURE[-1]
+            and self.max_decode_seq_len <= self.runner.max_seq_len_to_capture)
+
+        # If cuda graph can be used, pad tensors accordingly.
+        # See `capture_model` API for more details.
+        # vLLM uses cuda graph only for decoding requests.
+        cuda_graph_pad_size = 0
+        if use_captured_graph:
+            graph_batch_size = _get_graph_batch_size(batch_size)
+            assert graph_batch_size >= batch_size
+            cuda_graph_pad_size = graph_batch_size - batch_size
+            batch_size = graph_batch_size
+
+        # Tokens and positions.
+        self.input_tokens.extend([0] * cuda_graph_pad_size)
+        self.input_positions.extend([0] * cuda_graph_pad_size)
+        input_tokens_tensor = torch.tensor(self.input_tokens,
+                                           dtype=torch.long,
+                                           device=self.runner.device)
+        input_positions_tensor = torch.tensor(self.input_positions,
+                                              dtype=torch.long,
+                                              device=self.runner.device)
+
+        # Sequence and query lengths.
+        self.seq_lens.extend([1] * cuda_graph_pad_size)
+
+        # Attention metadata.
+        attn_metadata = self.attn_metadata_builder.build(
+            self.runner, self.seq_lens, self.query_lens, use_captured_graph,
+            cuda_graph_pad_size, batch_size)
```

Comment:
nit: should we remove `use_captured_graph ` arg and use cuda_graph_pad_size==-1 as a meaning that there's no cuda graph?

Reply:
also batch_size could be also inferred from query_lens?

Reply:
- Good idea. We could just use `cuda_graph_pad_size ==0` to infer that.
- Unfortunately batch size cannot be inferred from query lens because query lens were the original length without CUDA graph.